### ISSUE TITLE
Terathon::Arctan() too imprecise

### ISSUE THREAD

**Issue by DanielGibson** [EricLengyel/Terathon-Math-Library#3]
Hi,

We (Masterbrain Bytes/Cattle and Crops) have camera code that uses polar coordinates to handle input, and we also want the camera to move (and turn) with the vehicle the player is currently driving.
So we're using code like
```c++
float zrot = Arctan( vehicleTransform.m10, vehicleTransform.m00 );
azimuth += zrot;
```
This generally works, but we noticed that when just driving in a circle, at certain angles the camera stuttered.
When replacing `Arctan()` with `atan2f()` those problems went away.

This could've been it, but this issue somehow nerdsniped me so I started investigating and found out that `Arctan(float x)` has errors of up to `0.004459` (compared to `atan(x)` in double precision).

When looking up how a fast `atan()` can be implemented, I stumbled upon the idea of just doing a linear interpolation between values of a lookup table.
So I implemented it (with the existing lookup-table), just interpolating between `table[i]` and `table[i+1]`, using `b - i` as the lerp factor, and to my big surprise this works really well - the maximal error seems to be below `0.000020`, i.e. over 200 times lower.
In addition to that, it even seems to be about twice as fast, and it works with our camera code.

I tested this extensively both with values between +/- 10'000 (in steps of 0.00001) and values between +/- 1.5 (in steps of 0.00000001) and it indeed seems to work quite well.

Here's my test program (tested on both Linux and Windows), including both your and my `Arctan(float)` implementation: [atantest-upd.cpp.txt](https://github.com/EricLengyel/Terathon-Math-Library/files/5857213/atantest-upd.cpp.txt) (Github doesn't allow attaching .cpp files so I had to append .txt).
**UPDATE:** *Replaced the file with an updated one with one more element in table; also added function with 130 element lookup table.*
Feel free to use it (also in your commercial code).

Here's my test results from both Linux (on a AMD Ryzen 2700X) and Win10 (on an Intel Haswell i7-4771): [atanresults.txt](https://github.com/EricLengyel/Terathon-Math-Library/files/5853020/atanresults.txt)

Cheers,
Daniel

**Comment by DanielGibson** [comment]
I realized there was a little bug in my initial implementation: as `table[i+1]` is accessed, for `x == 1.0F` that's `table[65]` so it needs 66 elements instead of 65 to be correct.
I didn't notice that in my normal tests because `lerpfac == 0` in that case, so the value read from uninitialized memory doesn't influence the result.
Just for reference I added a third implementation (`Mod2_Terathon_Arctan()`) which is basically the same as the other, but uses a 130 element lookup table (=> multiplying `v` with `128` instead of `64`) - at least in my microbenchmark it's just as fast as `Mod_Terathon_Arctan()`, but the maximal error (and also the accumulated error) is about 4 times lower.

(I updated the linked implementation in the first post accordingly)

**Comment by EricLengyel** [comment]
Thanks for pointing this out. I looked into it and found that there was a bug in the `Arctan()` function. It should always have been able to do better than linear interpolation of the lookup values, but it was using an incorrect trig identity. I have updated the function with the correct identity, and the maximum error is now less than `2.0e-7`, which is 100 times smaller than plain lerp with the original table.
